File: scripts/BM.GPU.H100.8/gpu_clk_check.py
"""
# This script checks the GPU clock speed to determine if it is within the expected threshold of the maximum setting.
"""

import subprocess
import shlex
import json

# Function to run command and capture output
def run_cmd(cmd):
    cmd_split = shlex.split(cmd)
    try:
        results = subprocess.run(cmd_split, shell=False, stdout=subprocess.PIPE,
                                 stderr=subprocess.STDOUT, check=True, encoding='utf8')
        output = results.stdout.splitlines()
    except subprocess.CalledProcessError as e_process_error:
        return [f"Error: {cmd} {e_process_error.returncode} {e_process_error.output}"]
    return output

# Function to parse nvidia-smi output and emit JSON
def run_gpu_clk_check():
    """ Check if the clock speed is set to max """
    config = {
        "gpu_clk_check": {
            "state": "disable",
            "clock_speed": "1980"
        },
        "gpu_pci_ids": [
            "0000:0f:00.0",
            "0000:2d:00.0",
            "0000:44:00.0",
            "0000:5b:00.0",
            "0000:89:00.0",
            "0000:a8:00.0",
            "0000:c0:00.0",
            "0000:d8:00.0"
        ]
    }
    max_clock_speed = config["gpu_clk_check"]["clock_speed"]
    cmd = f'nvidia-smi --query-gpu=clocks.current.graphics --format=csv,noheader'
    output = run_cmd(cmd)
    result = parse_gpu_clk_results(output, max_clock_speed, len(config["gpu_pci_ids"]))
    return result


# Function to parse GPU clock to determine if is within the acceptable range
def parse_gpu_clk_results(results="undefined", max_clock_speed="undefined", gpu_count=8):
    result = {
        "gpu":
            {"max_clock_speed": "PASS"}
    }
    fail_list = []
    warn_list = []
    gpu = 0
    original_max_clock_speed = max_clock_speed
    allowed_max_clock_speed = ""
    if not results:
        result["gpu"]["max_clock_speed"] = "FAIL - check GPU "
    for line in results:
        if "couldn't communicate with the NVIDIA driver" in line:
            result["gpu"]["max_clock_speed"] = "FAIL - NVIDIA driver is not loaded"
            return result
        if "Error" in line:
            result["gpu"]["max_clock_speed"] = "FAIL - not able to run command 'nvidia-smi " \
                                               "--query-gpu=clocks.current.graphics --format=csv' "
            return result
    # Allow for an offset for H100
    max_clock_speed = str(int(original_max_clock_speed) - round(int(original_max_clock_speed) * 0.10))
    current_speed = ""
    allowed_speed = ""
    for line in results:
        if len(line) > 0:
            current_speed = line.split()[0]
            # Include the smallest allowed clock among all GPU clocks for allowed message
            if int(current_speed) >= int(max_clock_speed) and int(current_speed) < int(original_max_clock_speed):
                allowed_speed = str(min(int(allowed_speed), int(current_speed))) if allowed_speed else current_speed
            try:
                if int(current_speed) < int(max_clock_speed):
                    fail_list.append(str(gpu))
            except ValueError:
                fail_list.append(str(gpu))
        gpu += 1
    if fail_list:
        result["gpu"]["max_clock_speed"] = "FAIL - check GPU " + ",".join(iter(fail_list))
    else:
        if warn_list:
            result["gpu"]["max_clock_speed"] = f"WARN - Clock speed is lower than max clock speed. Check " + ", ".join(iter(warn_list))
        if results:
            if not allowed_speed:
                allowed_speed = current_speed
            result["gpu"]["max_clock_speed"] = f"PASS - Expected {original_max_clock_speed}, " f"allowed {allowed_speed}"
    return result

# Main function to call run_gpu_clk_check and parse the results
def main(argv=None):
    result = run_gpu_clk_check()
    print(json.dumps(result, indent=2))

# Run the main function
if __name__ == "__main__":
    main()

File: internal/level1_tests/gpu_clk_check.go
package level1_tests

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/oracle/oci-dr-hpc-v2/internal/executor"
	"github.com/oracle/oci-dr-hpc-v2/internal/logger"
	"github.com/oracle/oci-dr-hpc-v2/internal/reporter"
	"github.com/oracle/oci-dr-hpc-v2/internal/test_limits"
)

// GPUClkCheckTestConfig represents the config needed to run this test
type GPUClkCheckTestConfig struct {
	IsEnabled        bool   `json:"enabled"`
	Shape            string `json:"shape"`
	ExpectedClkSpeed int    `json:"clock_speed"`
}

// getGpuClkCheckTestConfig gets test config needed to run this test
func getGpuClkCheckTestConfig() (*GPUClkCheckTestConfig, error) {
	// Get shape from IMDS
	shape, err := executor.GetCurrentShape()
	if err != nil {
		return nil, err
	}

	// Load configuration from test_limits.json
	limits, err := test_limits.LoadTestLimits()
	if err != nil {
		return nil, err
	}

	// Initialize with defaults (H100 values - 1980 MHz as per Python script)
	gpuClkCheckTestConfig := &GPUClkCheckTestConfig{
		IsEnabled:        false,
		Shape:            shape,
		ExpectedClkSpeed: 1980, // Default H100 clock speed in MHz
	}

	// Check if test is enabled for this shape
	enabled, err := limits.IsTestEnabled(shape, "gpu_clk_check")
	if err != nil {
		return nil, err
	}
	gpuClkCheckTestConfig.IsEnabled = enabled

	// Get threshold configuration if available
	threshold, err := limits.GetThresholdForTest(shape, "gpu_clk_check")
	if err == nil {
		// Parse threshold configuration from test_limits.json
		switch v := threshold.(type) {
		case float64:
			gpuClkCheckTestConfig.ExpectedClkSpeed = int(v)
		case map[string]interface{}:
			if clockSpeed, ok := v["clock_speed"].(float64); ok {
				gpuClkCheckTestConfig.ExpectedClkSpeed = int(clockSpeed)
			}
		}
	}

	return gpuClkCheckTestConfig, nil
}

// getGPUClockSpeeds uses nvidia-smi to get current GPU clock speeds
func getGPUClockSpeeds() ([]string, error) {
	// Use nvidia-smi to query current graphics clock speeds
	result := executor.RunNvidiaSMIQuery("clocks.current.graphics")
	if !result.Available {
		return nil, fmt.Errorf("nvidia-smi not available: %s", result.Error)
	}

	// Check for driver communication issues
	if strings.Contains(result.Output, "couldn't communicate with the NVIDIA driver") {
		return nil, fmt.Errorf("NVIDIA driver is not loaded")
	}

	// Parse the output
	output := strings.TrimSpace(result.Output)
	if output == "" {
		return []string{}, nil
	}

	// Split by lines and collect clock speeds
	lines := strings.Split(output, "\n")
	var clockSpeeds []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			clockSpeeds = append(clockSpeeds, line)
		}
	}

	return clockSpeeds, nil
}

// validateGPUClockSpeeds validates clock speeds against expected threshold
// Following Python logic: allow 90% of expected speed (10% tolerance)
func validateGPUClockSpeeds(clockSpeeds []string, expectedSpeed int) (string, string, error) {
	if len(clockSpeeds) == 0 {
		return "FAIL", "check GPU", fmt.Errorf("no GPU clock speeds found")
	}

	// Calculate minimum acceptable speed (90% of expected - 10% tolerance for H100)
	minAcceptableSpeed := expectedSpeed - int(float64(expectedSpeed)*0.10)
	
	var failedGPUs []string
	var allowedSpeed string
	var minAllowedSpeed int = -1
	
	for gpuIndex, speedStr := range clockSpeeds {
		// Extract numeric value from speed string (e.g., "1980 MHz" -> "1980")
		fields := strings.Fields(speedStr)
		if len(fields) == 0 {
			failedGPUs = append(failedGPUs, strconv.Itoa(gpuIndex))
			continue
		}
		
		currentSpeedStr := fields[0]
		currentSpeed, err := strconv.Atoi(currentSpeedStr)
		if err != nil {
			failedGPUs = append(failedGPUs, strconv.Itoa(gpuIndex))
			continue
		}
		
		// Check if speed is below minimum threshold
		if currentSpeed < minAcceptableSpeed {
			failedGPUs = append(failedGPUs, strconv.Itoa(gpuIndex))
		} else {
			// Speed is acceptable - track the minimum for reporting
			if minAllowedSpeed == -1 || currentSpeed < minAllowedSpeed {
				minAllowedSpeed = currentSpeed
				allowedSpeed = currentSpeedStr
			}
		}
	}

	if len(failedGPUs) > 0 {
		return "FAIL", "check GPU " + strings.Join(failedGPUs, ","), 
			fmt.Errorf("GPU clock speeds below threshold for GPUs: %s", strings.Join(failedGPUs, ","))
	}

	// All GPUs passed - format success message following Python pattern
	if allowedSpeed == "" && len(clockSpeeds) > 0 {
		// Use the last processed speed if no specific allowed speed was tracked
		fields := strings.Fields(clockSpeeds[len(clockSpeeds)-1])
		if len(fields) > 0 {
			allowedSpeed = fields[0]
		}
	}
	
	statusMsg := fmt.Sprintf("Expected %d, allowed %s", expectedSpeed, allowedSpeed)
	return "PASS", statusMsg, nil
}

func RunGPUClkCheck() error {
	logger.Info("=== GPU Clock Speed Check ===")
	testConfig, err := getGpuClkCheckTestConfig()
	if err != nil {
		return err
	}

	if !testConfig.IsEnabled {
		errorStatement := fmt.Sprintf("Test not applicable for this shape %s", testConfig.Shape)
		logger.Info(errorStatement)
		return errors.New(errorStatement)
	}

	logger.Info("Starting GPU clock speed check...")
	rep := reporter.GetReporter()

	// Step 1: Get GPU clock speeds
	logger.Info("Step 1: Getting GPU clock speeds...")
	clockSpeeds, err := getGPUClockSpeeds()
	if err != nil {
		logger.Error("GPU Clock Check: FAIL - Could not get GPU clock speeds:", err)
		rep.AddGPUClockResult("FAIL", "", err)
		return fmt.Errorf("could not get GPU clock speeds: %w", err)
	}

	logger.Info("Found GPU clock speeds:", clockSpeeds)

	// Step 2: Validate clock speeds
	logger.Info("Step 2: Validating clock speeds...")
	logger.Info("Expected clock speed (MHz):", testConfig.ExpectedClkSpeed)

	status, statusMsg, validationErr := validateGPUClockSpeeds(clockSpeeds, testConfig.ExpectedClkSpeed)

	switch status {
	case "PASS":
		logger.Info("GPU Clock Check: PASS -", statusMsg)
		rep.AddGPUClockResult("PASS", statusMsg, nil)
		return nil
	default: // FAIL
		logger.Error("GPU Clock Check: FAIL -", statusMsg)
		rep.AddGPUClockResult("FAIL", statusMsg, validationErr)
		return validationErr
	}
}

func PrintGPUClkCheck() {
	logger.Info("GPU Clock Check: Checking GPU clock speeds...")
	logger.Info("GPU Clock Check: PASS - Clock speeds are within acceptable range")
}